Added a style guide for iris pytest (#5785)
SciTools/iris: docs/src/developers_guide/testing_style_guide.rst

Reviewer: Cross-reference to the pytest documentation whenever a concept is introduced. So perhaps most of the headings could be links.

E.g.

- Parameters: https://docs.pytest.org/en/stable/example/parametrize.html
- Parametrizing fixtures: https://docs.pytest.org/en/stable/how-to/fixtures.html#fixture-parametrize
- Etcetera

Reviewer: We're currently expecting people to read several different pages before they have a full picture:

https://github.com/SciTools/iris/blob/9a4a45b874501c2ace720628c7cde9afbd438dc3/docs/src/developers_guide/testing_style_guide.rst?plain=1#L5-L6

https://github.com/SciTools/iris/blob/9a4a45b874501c2ace720628c7cde9afbd438dc3/docs/src/developers_guide/contributing_testing.rst?plain=1#L6-L7

https://github.com/SciTools/iris/blob/9a4a45b874501c2ace720628c7cde9afbd438dc3/docs/src/developers_guide/testing_tools.rst?plain=1#L5-L6

- A confusing read
- Too easy for inconsistencies to creep in (several of which I have mentioned in this review)

For me this would be much better as a single **Writing Tests** page.

.. include:: ../common_links.inc

.. _testing_style_guide:

PyTest Style Guide
******************

This style guide should be approached pragmatically. Most of the guidelines laid out
below will not be practical in every scenario, and as such should not be considered
firm rules.

At time of writing, some existing tests have already been written in PyTest,
so might not be abiding by these guidelines.

Directory
=========

Where suitable, tests should be located within the relevant directories.
In most circumstance, that means new tests should not be placed in the
root test directory, but in the relevant sub-folders.

Conftest.py
===========

There should be a conftest.py file in the root/unit and root/integration

> folders. Additional lower level conftests can be added if it is agreed there

Reviewer: ```suggestion
folders. Additional lower level ``conftest``s can be added if it is agreed there
```

Maybe? For consistency and to draw attention?

is a need.

Fixtures
========

As far as is possible, the actual test function should do little else but the

> actual assertion. However, in some cases this will not be applicable, so this
> will have to be decided on a case by case basis.

Reviewer: I'd prefer less vague phrasing. You want to avoid stuff like:

>This is true except when it is not

You've done a great job of more specific phrasing in the following paragraph. So how to do the same with this one... The conflict here is readability versus maintainability:

- Max readability = the whole sequence written in a single test file - you can see everything in one place
- Max maintainability = all setup abstracted into fixtures - you can alter all relevant tests with a single change, distinction is made between setup and test when things go wrong

Something like the below?

```suggestion
actual assertion. Separating preparation into fixtures can make the code
harder to follow, so compromises are acceptable; e.g. setting up a test ``Cube``
should be a fixture, creating a simple string (``expected = "foo"``) for a single
test should not be a fixture. 
```

> New fixtures should always be considered for conftest when added. If it is

Reviewer: ```suggestion
New fixtures should always be considered for ``conftest`` when added. If it is
```

?

decided that they are not suitably reusable, they can be placed within the
local test file.

Parameterisation
================

Though it is a useful tool, we should not be complicating tests to work around
parameters; they should only be used when it is simple and apparent to implement.

Where you are parameterising multiple tests with the same parameters, it is
usually prudent to use the parameterisation within fixtures. When doing this,
ensure within the tests that it is apparent it's being parameterised,
either within the fixture name or with comments.

> All parameterisation benefits from ids, and so should be used where possible.

Reviewer: ```suggestion
All parameterisation benefits from ``ids``, and so should be used where possible.
```

Definitely deserves a link to provide context.

Classes
=======

How and when to group tests within classes can be based on personal opinion,

> we do not deem consistency on this a vital concern.

Reviewer: The below will need changing otherwise we had a direct contradiction.

https://github.com/SciTools/iris/blob/9a4a45b874501c2ace720628c7cde9afbd438dc3/docs/src/developers_guide/contributing_testing.rst?plain=1#L48-L52

https://github.com/SciTools/iris/blob/9a4a45b874501c2ace720628c7cde9afbd438dc3/docs/src/developers_guide/contributing_testing.rst?plain=1#L98-L101

Mocks
=====

> Any mocking should be done with pytest.mock, and monkeypatching where suitable.

Reviewer: ```suggestion
Any mocking should be done with ``pytest.mock``, and monkeypatching where suitable.
```

Need some more links. Could maybe consider [intersphinx](https://github.com/SciTools/iris/blob/9a4a45b874501c2ace720628c7cde9afbd438dc3/docs/src/conf.py#L236-L245) so you could use a format like `:func:...`, but that might be overkill.

.. include:: ../common_links.inc

.. _testing_style_guide:

PyTest Style Guide
******************

This style guide should be approached pragmatically. Most of the guidelines laid out
below will not be practical in every scenario, and as such should not be considered
firm rules.

At time of writing, some existing tests have already been written in PyTest,
so might not be abiding by these guidelines.

Directory
=========

Where suitable, tests should be located within the relevant directories.
In most circumstance, that means new tests should not be placed in the
root test directory, but in the relevant sub-folders.

> `conftest.py <https://docs.pytest.org/en/7.1.x/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files>`_

Reviewer: ```suggestion
`conftest.py <https://docs.pytest.org/en/stable/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files>`_
```
